mixer/pkg/attribute/envoyProtoBag.go

```diff
@@ -0,0 +1,328 @@
+	New: func() interface{} {
+		return &EnvoyProtoBag{
+			referencedAttrs: make(map[attr.Reference]attr.Presence, referencedAttrsSize),
+		}
+	},
+}
+
+func fillAddress(reqMap map[string]interface{}, address *core.Address, name string) {
+	reqMap[name+".ip"] = []byte(net.ParseIP(address.GetSocketAddress().GetAddress()).To16())
+	reqMap[name+".port"] = int64(address.GetSocketAddress().GetPortValue())
+}
+
+func reformatTime(theTime *timestamp.Timestamp, durationNanos int32) time.Time {
+	return time.Unix(theTime.Seconds, int64(theTime.Nanos)+int64(durationNanos))
+}
+
+
+func isGrpc(headers map[string]string) bool {
+ 	if val, ok := headers["content-type"]; ok {
+		return strings.Contains(val, "grpc")
+	}
+	return false
+}
+
+func fillContextProtocol(reqMap map[string]interface{}) {
+	if isGrpc(reqMap["request.headers"].(map[string]string)) {
+		reqMap["context.protocol"] = "grpc"
+	} else {
+		reqMap["context.protocol"] = "http"
+	}
+}
+
+// AuthzProtoBag returns an attribute bag for an Ext-Authz Check Request.
+// When you are done using the proto bag, call the Done method to recycle it.
+func AuthzProtoBag(req *authzGRPC.CheckRequest) *EnvoyProtoBag {
+	pb := envoyProtoBags.Get().(*EnvoyProtoBag)
+
+	// build the message-level dictionary
+	reqMap := make(map[string]interface{})
+
+	reqMap["context.reporter.kind"] = "inbound"
+	fillAddress(reqMap, req.GetAttributes().GetDestination().GetAddress(), "destination")
+	fillAddress(reqMap, req.GetAttributes().GetSource().GetAddress(), "source")
+	reqMap["destination.principal"] = req.GetAttributes().GetDestination().GetPrincipal()
+	reqMap["source.principal"] = req.GetAttributes().GetSource().GetPrincipal()
+	reqMap["request.headers"] = attr.WrapStringMap(req.GetAttributes().GetRequest().GetHttp().GetHeaders())
+	reqMap["request.host"] = req.GetAttributes().GetRequest().GetHttp().GetHost()
+	reqMap["request.method"] = req.GetAttributes().GetRequest().GetHttp().GetMethod()
+	reqMap["request.path"] = req.GetAttributes().GetRequest().GetHttp().GetPath()
+	reqMap["request.scheme"] = req.GetAttributes().GetRequest().GetHttp().GetScheme()
+	reqMap["request.time"] = reformatTime(req.GetAttributes().GetRequest().GetTime(), 0)
+	reqMap["request.useragent"] = req.GetAttributes().GetRequest().GetHttp().GetHeaders()["user-agent"]
+	fillContextProtocol(reqMap)
+	pb.reqMap = reqMap
+
+	scope.Debugf("Returning bag with attributes:\n%v", pb)
+
+	return pb
+}
+
+// AccessLogProtoBag returns an attribute bag from a StreamAccessLogsMessage
+// When you are done using the proto bag, call the Done method to recycle it.
+// num is the index of the entry from the message's batch to create a bag from
+func AccessLogProtoBag(msg *accessLogGRPC.StreamAccessLogsMessage, num int) *EnvoyProtoBag {
+	// build the message-level dictionary
+	pb := envoyProtoBags.Get().(*EnvoyProtoBag)
+	reqMap := make(map[string]interface{})
+	if httpLogs := msg.GetHttpLogs(); httpLogs != nil {
+		reqMap["context.reporter.kind"] = "inbound"
```

Comment:
@douglas-reid Do you think if it is ok that we listed grpcAccessLog shim is only available for inbound? I cannot think of a way to distinguish outbound and inbound.

Reply:
I think that is fine -- by their very nature, I would expect access logs to be inbound only. This will represent a loss in some reporting features, but I'm not sure that there is much we can do about that.